### ISSUE TITLE
Secure redis?

### ISSUE THREAD

**Issue by ninja-** [sameersbn/docker-gitlab#619]
I see that while the postgres database can be password protected, redis is exposed to every process on the same machine. Do you think there's an easy solution for docker-compose usage?

**Comment by outcoldman** [comment]
@ninja- using docker networking?

**Comment by lenovouser** [comment]
This is a duplicate of #374 and currently not supported by GitLab. I have created an issue for that **[here](https://gitlab.com/gitlab-org/gitlab-ce/issues/13305)**.

**Comment by ninja-** [comment]
@outcoldman yeah using docker networking. I can't find a way to isolate redis within docker network while exposing gitlab.

**Comment by frenchbeard** [comment]
@ninja- @outcoldman I don't know the extent of isolation it provides, but I compartimented the 3 containers in a "back-gitlab" network, put an nginx reverse-proxy (in its own container) in this network and a "front" network, only exposing the reverse proxy. Hope this provides you with at least an idea.

**Comment by ninja-** [comment]
@frenchbeard the problem is in making gitlab be both on internal and external network :/

**Comment by outcoldman** [comment]
@ninja- why is this a problem? you can use [docker network connect](https://docs.docker.com/engine/reference/commandline/network_connect/) for that. Just start gitlab in `gitlab` network and add it later to `front` network using `docker network connect`

**Comment by sameersbn** [comment]
I will check if we can use authentication password with redis.

**Comment by frenchbeard** [comment]
@ninja- @outcoldman I know in my case I simply use the docker-compose file v2 syntax to specify it. I don't know which would be best though, having the reverse-proxy container simply linked to the gitlab container or plug it on the "back" network ?

**Comment by outcoldman** [comment]
@frenchbeard it really depends on the environment and how secure you want to be. You can always go with the third option - have a dedicated network between proxy and gitlab.

**Comment by ninja-** [comment]
@frenchbeard could you share your compose config? If it's really that easy we can add it to official config

**Comment by frenchbeard** [comment]
@ninja- I have a big compose file for my server (with a lot of things tied together), I'll work on a simple one regarding only gitlab and nginx. It'll still be tied to jwilder's [automated proxy](https://github.com/jwilder/nginx-proxy) though. I'll come back to you asap, tomorrow if decoupling things does not break the configuration.

**Comment by frenchbeard** [comment]
Sorry for the delay, I kind of broke my config as I went on producing the Compose file.
The one thing problematic with v2 is that linked containers no longer provide the environment environment variables you use to connect to both Redis and Postgresql containers (IP and PORT), meaning we'll have to either change the way you get this information within the gitlab container or provide it directly in the Compose file (which I chose to do).
I'm also trying to include the letsencrypt bit from #594 in this.

**Comment by frenchbeard** [comment]
I left out the letsencrypt part for now, since testing it too many times ends up getting you locked out of the automated process. You can check the configuration file on my [fork](https://github.com/frenchbeard/docker-gitlab), with a somewhat secure nginx template file provided for the [docker-gen](https://github.com/jwilder/docker-gen) container.

**Comment by ninja-** [comment]
That's close. Would it be enough to put redis in network backend and gitlab in networks backend and frontend?

**Comment by frenchbeard** [comment]
That would do, and the same thing can be done for the postgresql, you could base it on the PR regarding switching the compose file to v2 #631, only adding the following :
```yaml
version: '2'
services:
   postgresql:
[...]
      networks:
         - backend
[...]
   redis:
[...]
      networks:
         - backend
[...]
   gitlab:
[...]
      networks:
         - backend
         - frontend
[...]
networks:
   frontend:
      driver: bridge
   backend:
      driver: bridge
```
